### PR TITLE
DM-38403: Add calspec/simbad tables to local cache

### PR DESCRIPTION
**Pull request by mfisherlevine**

### CODE REVIEW

File: spectractor/tools.py [hunk 1]
```diff
@@ -2469,6 +2472,50 @@ def uvspec_available():
...
+    Returns
+    -------
+    success : `bool`
+        Were the necessary file(s) successfully obtained?
+    """
+
+    def patchSimbadURL(simbad):
+        """Monkeypatch the URL that Simbad is using to force it to use https.
+        """
+        simbad.SIMBAD_URL = (simbad.SIMBAD_URL.replace('http', 'https')
```
Comment (ktlim) [inline spectractor/tools.py]: Isn't this simpler (and slightly safer) as `simbad.SIMBAD_URL = simbad.SIMBAD_URL.replace("http:", "https:")` ?  You don't need the whole `if` clause in that case.

File: spectractor/tools.py [hunk 2]
```diff
@@ -2469,6 +2472,50 @@ def uvspec_available():
...
+    file, it should take the cached copy, and the test should pass.
+
+    Returns
+    -------
+    success : `bool`
+        Were the necessary file(s) successfully obtained?
+    """
+
+    def patchSimbadURL(simbad):
+        """Monkeypatch the URL that Simbad is using to force it to use https.
```
Comment (ktlim) [inline spectractor/tools.py]: Would be nice to say why this is necessary, maybe especially since it seems that the reason is lost in time...

**Review by ktlim — APPROVED** [review]

File: spectractor/tools.py (hunk 2 above)
Comment (timj) [inline spectractor/tools.py]: Is this the actual problem? ie with this patch everything works? If so, maybe there should be a issue filed with Astroquery?

File: spectractor/tools.py [hunk 3]
```diff
@@ -2469,6 +2472,50 @@ def uvspec_available():
...
+
+    for target in ['HD 111980']:
+        try:
+            calspec = getCalspec.Calspec(target)
+            calspec.get_spectrum_numpy()
+        except (RuntimeError, requests.exceptions.HTTPError):
+            return False
+
+        try:
+            simbadQuerier = SimbadClass()
```
Comment (timj) [inline spectractor/tools.py]: Do these two lines need to be inside the loop?

**Review by timj — APPROVED** [review]
Looks okay. Kind of cute, but looks okay 😄

File: spectractor/tools.py (hunk 3 above)
Comment (ktlim) [reply spectractor/tools.py]: In fact this whole `try` block looks like it doesn't depend on `target`.

Comment (timj) [reply spectractor/tools.py]: I think the `astroquery_label` comes from `calspec` variable which depends on target.

File: tests/test_extractor.py [hunk 4]
```diff
@@ -1,6 +1,10 @@
 import matplotlib as mpl
 mpl.use('Agg')  # must be run first! But therefore requires noqa E402 on all other imports
 
+# This must be run before any astropy importing.
```
Comment (timj) [inline tests/test_extractor.py]: Does this comment rely on pytest loading test files in the right order?

File: tests/data/cache/astropy/download/url/e9b5d99b7f6684fbea4a485dc468b804/url [hunk 5]
```diff
@@ -0,0 +1 @@
+https://archive.stsci.edu/hlsps/reference-atlases/cdbs/calspec/hd111980_stis_005.fits
```
Comment (timj) [inline tests/data/cache/astropy/download/url/e9b5d99b7f6684fbea4a485dc468b804/url:1]: The fix is to permanently cache the file and so never go to the internet?

File: tests/test_extractor.py (hunk 4 above)
Comment (erykoff) [reply tests/test_extractor.py]: First, it's not pytest here, it's nose.  Second, this would be a separate process, right?  So I think that's what matters.

File: tests/data/cache/astropy/download/url/e9b5d99b7f6684fbea4a485dc468b804/url (hunk 5 above)
Comment (erykoff) [reply tests/data/cache/astropy/download/url/e9b5d99b7f6684fbea4a485dc468b804/url:1]: Yup!  So the tests can run happily no matter what.

File: spectractor/tools.py [hunk 6]
```diff
@@ -10,12 +10,15 @@
 from astropy.stats import sigma_clip, sigma_clipped_stats
 from astropy.io import fits
 from astropy import wcs as WCS
+from getCalspec import getCalspec
```
Comment (timj) [inline spectractor/tools.py]: Are the changes in tools.py needed any more? Doesn't flake8 complain?

File: tests/test_extractor.py (hunk 4 above)
Comment (timj) [reply tests/test_extractor.py]: Ok. For pytest if you tell it to load 3 test files it will load them in the order you give to pytest and if the first test file imports astropy the comment in the second file about setting the env var won't do anything. In xdist mode you are at the mercy of the subprocess import order as well. Does astropy not have a way to programmatically add the cache dir?

File: spectractor/tools.py (hunk 6 above)
Comment (erykoff) [reply spectractor/tools.py]: You're right, but oh spectractor is not flake8 safe.

File: spectractor/tools.py [hunk 7]
```diff
@@ -16,6 +16,7 @@
 from matplotlib.ticker import MaxNLocator
 
 import warnings
+import requests
```
Comment (timj) [inline spectractor/tools.py]: Also this line...

File: tests/test_fullchain.py [hunk 8]
```diff
@@ -11,14 +11,17 @@
 from spectractor.logbook import LogBook  # noqa: E402
 from spectractor.config import load_config  # noqa: E402
 from spectractor.simulation.image_simulation import ImageSim  # noqa: E402
-from spectractor.tools import plot_spectrum_simple, uvspec_available  # noqa: E402
+from spectractor.tools import (plot_spectrum_simple,  # noqa: E402
```
Comment (timj) [inline tests/test_fullchain.py]: I think you can revert this change as well.

**Review by timj — APPROVED** [review]
The caching and the temporary cache decorator look good to me. Much better solution.
